src/java/org/apache/cassandra/cql3/statements/SelectStatement.java

```diff
@@ -405,17 +405,20 @@ public ReadQuery getQuery(QueryOptions options,
                               long nowInSec,
                               DataLimits limit)
     {
-        boolean isPartitionRangeQuery = restrictions.isKeyRange() || restrictions.usesSecondaryIndexing();
+        RowFilter rowFilter = getRowFilter(options, state);
 
-        if (isPartitionRangeQuery)
+        if (restrictions.isKeyRange())
         {
-            if (restrictions.isKeyRange() && restrictions.usesSecondaryIndexing() && !SchemaConstants.isLocalSystemKeyspace(table.keyspace))
+            if (restrictions.usesSecondaryIndexing() && !SchemaConstants.isLocalSystemKeyspace(table.keyspace))
                 Guardrails.nonPartitionRestrictedIndexQueryEnabled.ensureEnabled(state);
 
-            return getRangeCommand(options, state, columnFilter, limit, nowInSec);
+            return getRangeCommand(options, state, columnFilter, rowFilter, limit, nowInSec);
         }
 
-        return getSliceCommands(options, state, columnFilter, limit, nowInSec);
+        if (restrictions.usesSecondaryIndexing() && !rowFilter.isStrict())
```

Comment:
If we aren't able to do strict filtering, we just revert back to doing a range read. If we didn't do this, we'd have situations where we don't get a digest match downstream, but still need to apply post-filtering, which introduces possible bugs w/ short-reads, and I didn't want to complicate this patch with all that. In the end, the logic here is the same as before the patch with the `restrictions.usesSecondaryIndexing()` check pulled out of the first conditional above and the `isStrict()` check.

Filtering queries that don't use an index are still horribly broken at CL > ONE, but I'm not trying to fix [CASSANDRA-19007](https://issues.apache.org/jira/browse/CASSANDRA-19007) here.